src/figuro/renderer/openglSiwin.nim

```diff
@@ -51,77 +52,65 @@ proc setupWindow*(
     )
 
-  window.makeContextCurrent()
-
   let winCfg = frame.loadLastWindow()
 
-  window.`style=`(style)
-  window.`pos=`(winCfg.pos)
-
 proc newSiwinRenderer*(
     frame: WeakRef[AppFrame],
     forcePixelScale: float32,
     atlasSize: int,
 ): RendererSiwin =
-  let window = newWindow("Figuro", ivec2(1280, 800), visible = false)
-  result = RendererSiwin(window: window, frame: frame)
-  startOpenGL(openglVersion)
+  info "starting siwin renderer"
+  let globals = newSiwinGlobals(
+    preferedPlatform =
+      case getEnv("FIGURO_SIWIN_BACKEND", "auto")
+      of "x11": x11
+      of "wayland": wayland
+      else: defaultPreferedPlatform()
+  )
+
+  let window = newOpenglWindow(globals, title = "Figuro", size = ivec2(1280, 800))
+  var renderer = RendererSiwin(window: window, frame: frame, atlasSize: atlasSize)
 
+  startOpenGL(openglVersion)
   setupWindow(frame, window)
 
-  configureRenderer(result, frame, forcePixelScale, atlasSize)
-
-proc convertStyle*(fs: FrameStyle): WindowStyle =
-  case fs
-  of FrameStyle.DecoratedResizable:
-    WindowStyle.DecoratedResizable
-  of FrameStyle.DecoratedFixedSized:
-    WindowStyle.Decorated
-  of FrameStyle.Undecorated:
-    WindowStyle.Undecorated
-  of FrameStyle.Transparent:
-    WindowStyle.Transparent
-
-proc toUi*(wbtn: windex.ButtonView): set[UiMouse] =
-  when defined(nimscript):
-    for b in set[Button](wbtn):
-      result.incl UiButton(b.int)
-  else:
-    copyMem(addr result, unsafeAddr wbtn, sizeof(ButtonView))
+  configureRenderer(renderer, frame, forcePixelScale, atlasSize, initializeContext = false)
+
+  renderer
 
 method swapBuffers*(r: RendererSiwin) =
-  r.window.swapBuffers()
+  # r.window.swapBuffers()
+  # It's a no-op for now.
+  return
 
 method pollEvents*(r: RendererSiwin) =
-  windex.pollEvents()
+  # It's a no-op on this backend.
+  return
 
 method getScaleInfo*(r: RendererSiwin): ScaleInfo =
-  let scale = r.window.contentScale()
-  result.x = scale
-  result.y = scale
-
-var lastMouse = Mouse()
-
-proc copyInputs*(w: Window): AppInputs =
-  result = AppInputs(mouse: lastMouse)
-  result.buttonRelease = toUi w.buttonReleased()
-  result.buttonPress = toUi w.buttonPressed()
-  result.buttonDown = toUi w.buttonDown()
-  result.buttonToggle = toUi w.buttonToggle()
-
-method copyInputs*(r: RendererSiwin): AppInputs =
-  copyInputs(r.window)
+  # TODO: implement
+  result.x = 1
+  result.y = 1
 
 method setClipboard*(r: RendererSiwin, cb: ClipboardContents) =
-  match cb:
-    ClipboardStr(str):
-      windex.setClipboardString(str)
-    ClipboardEmpty:
-      discard
+  warn "TODO: siwin backend: clipboard write"
+  return
 
 method getClipboard*(r: RendererSiwin): ClipboardContents =
-  let str = windex.getClipboardString()
-  return ClipboardStr(str)
+  warn "TODO: siwin backend: clipboard read"
+  return ClipboardStr("")
+
+method run*(renderer: RendererSiwin) =
+  renderer.window.run(
```

Comment:
This will break things as the Figuro Renderer needs to control things and poll for events. It has it's own event system which controls when rendering needs to occur.

Windy/Windex had a `poll` and exposes a `swapBuffer`. It looks like Siwin doesn't expose a poll method, but has some `swapBuffer` methods.

I'm going to create an issue for Siwin about it.

Reply:
Probably what would be needed on the Figuro side is to change the `runRendererLoop` or maybe the `pollAndRender`. Something like this:

```nim

proc runRendererLoop*(renderer: Renderer) =
  threadEffects:
    RenderThread

  siwinWindw.eventHandler.onRender = proc () =
    renderer.pollAndRender()

  siwinWindow.firstStep() # first step

  while app.running:
    siwinWindow.redraw() # request redraw
    siwinWindow.step()

    os.sleep(renderer.duration.inMilliseconds)

redraw

```